contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol

```diff
@@ -0,0 +1,437 @@
+
+        // Fill the orders.
+        uint256 singleProtocolFee = exchange.protocolFeeMultiplier().safeMul(tx.gasprice);
+        uint256 ethRemaining = address(this).balance;
+        uint256 boughtAmount = 0;
+        uint256 soldAmount = 0;
+        for (uint256 i = 0; i < data.orders.length; ++i) {
+            // Check if we've hit our targets.
+            if (data.side == Side.Sell) {
+                // Market sell check.
+                if (soldAmount >= data.fillAmount) {
+                    break;
+                }
+            } else {
+                // Market buy check.
+                if (boughtAmount >= data.fillAmount) {
+                    break;
+                }
+            }
+
+            // Ensure we have enough ETH to cover the protocol fee.
+            if (ethRemaining < singleProtocolFee) {
+                LibTransformERC20RichErrors
+                    .InsufficientProtocolFeeError(ethRemaining, singleProtocolFee)
+                    .rrevert();
+            }
+
+            // Fill the order.
+            FillOrderResults memory results;
+            if (data.side == Side.Sell) {
+                // Market sell.
+                results = _sellToOrder(
+                    data.buyToken,
+                    data.sellToken,
+                    data.orders[i],
+                    data.signatures[i],
+                    data.fillAmount.safeSub(soldAmount).min256(
+                        data.maxOrderFillAmounts.length > i
+                        ? data.maxOrderFillAmounts[i]
+                        : uint256(-1)
+                    ),
+                    singleProtocolFee
+                );
+            } else {
+                // Market buy.
+                results = _buyFromOrder(
+                    data.buyToken,
+                    data.sellToken,
+                    data.orders[i],
+                    data.signatures[i],
+                    data.fillAmount.safeSub(boughtAmount).min256(
+                        data.maxOrderFillAmounts.length > i
+                        ? data.maxOrderFillAmounts[i]
+                        : uint256(-1)
+                    ),
+                    singleProtocolFee
+                );
+            }
+
+            // Accumulate totals.
+            soldAmount = soldAmount.safeAdd(results.takerTokenSoldAmount);
+            boughtAmount = boughtAmount.safeAdd(results.makerTokenBoughtAmount);
+            ethRemaining = ethRemaining.safeSub(results.protocolFeePaid);
+        }
+
+        // Ensure we hit our targets.
+        if (data.side == Side.Sell) {
+            // Market sell check.
+            if (soldAmount < data.fillAmount) {
+                LibTransformERC20RichErrors
+                    .IncompleteFillSellQuoteError(
+                        address(data.sellToken),
+                        soldAmount,
+                        data.fillAmount
+                    ).rrevert();
+            }
+        } else {
```

Comment:
Is there an automatic check when decoding that enforces `data.side` is `Side.Sell` or `Side.Buy`? If not we should probably do `...else if (data.side == Side.Buy) {..} else {revert()}`

Reply:
Yeah, the `abi.decode()` will do a range check on the enum value.